mountpoint-s3/tests/fuse_tests/prefetch_test.rs

```diff
@@ -0,0 +1,101 @@
+use fuser::BackgroundSession;
+use mountpoint_s3::prefetch::PrefetcherConfig;
+use mountpoint_s3::S3FilesystemConfig;
+use std::fs::OpenOptions;
+use std::io::Read;
+use std::io::Seek;
+use tempfile::TempDir;
+use test_case::test_case;
+
+use crate::fuse_tests::TestClientBox;
+
+// test for checking either prefetching fails or read original object when object is mutated during read.
+fn prefetch_test_etag<F>(creator_fn: F, prefix: &str, request_size: usize, read_size: usize)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    const OBJECT_SIZE: usize = 1024 * 1024;
+    const SEEK_POS: u64 = 200 * 1024;
+
+    let prefetcher_config = PrefetcherConfig {
+        first_request_size: request_size,
+        ..Default::default()
+    };
+
+    let test_config = S3FilesystemConfig {
+        prefetcher_config,
+        ..Default::default()
+    };
+
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);
+    let mut read_buf = vec![0u8; OBJECT_SIZE];
+
+    // Make sure there's an existing directory
+    test_client.put_object("dir/hello.txt", &read_buf).unwrap();
+
+    let mut path = mount_point.path().join("dir/hello.txt");
+
+    // Opening the file to read and write
+    let mut f = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .expect("should be able to open the file");
+
+    let mut buf = vec![0u8; read_size];
+    f.read_exact(&mut buf).expect("Should be able to read file to buf");
+
+    read_buf = vec![255u8; OBJECT_SIZE];
+    test_client.put_object("dir/hello.txt", &read_buf).unwrap();
+
+    // In order to read from a position where prefetching next block is necessary
+    f.seek(std::io::SeekFrom::Start(SEEK_POS))
+        .expect("Seek position should be less than object size");
+    let mut dest_buf = vec![0u8; read_size];
+    match f.read_exact(&mut dest_buf) {
+        Ok(()) => assert_eq!(buf, dest_buf),
+        Err(err) => assert_eq!(err.raw_os_error(), Some(libc::EIO)),
+    };
+
+    drop(f);
+
+    // File wont be mutated during read now to check if prefetcher reads correctly otherwise.
```

Comment:
I think the actual explanation here is that re-opening the file will see the new file contents.